web-api/src/storage/postgres.rs

```diff
@@ -284,89 +297,158 @@ impl storage::Interest for Storage {
+                    view_count = EXCLUDED.view_count,
+                    view_time_ms = EXCLUDED.view_time_ms,
+                    last_view = EXCLUDED.last_view;",
+                )
+                .build()
+                .execute(&mut *tx)
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    async fn upsert_interactions<'d, I>(
+        tx: &mut Transaction<'_, Postgres>,
+        user_id: &UserId,
+        now: DateTime<Utc>,
+        interactions: I,
     ) -> Result<(), Error>
     where
-        F: Fn(&mut Vec<PositiveCoi>) -> &PositiveCoi + Send + Sync,
+        I: IntoIterator<Item = (&'d DocumentId, UserInteractionType)>,
+        I::IntoIter: ExactSizeIterator,
     {
-        let mut tx = self.postgres.pool.begin().await?;
-
-        sqlx::query("INSERT INTO coi_update_lock (user_id) VALUES ($1) ON CONFLICT DO NOTHING;")
-            .bind(user_id)
-            .execute(&mut tx)
-            .await?;
-        sqlx::query("SELECT FROM coi_update_lock WHERE user_id = $1 FOR UPDATE;")
-            .bind(user_id)
-            .execute(&mut tx)
-            .await?;
-
-        // fine as we convert it to i32 when we store it in the database
-        #[allow(clippy::cast_sign_loss)]
-        let mut positive_cois: Vec<_> = sqlx::query_as::<_, QueriedCoi>(
-            "SELECT coi_id, is_positive, embedding, view_count, view_time_ms, last_view
-            FROM center_of_interest
-            WHERE user_id = $1 AND is_positive;",
-        )
-        .bind(user_id)
-        .fetch_all(&mut tx)
-        .await?
-        .into_iter()
-        .map(|coi| PositiveCoi {
-            id: coi.coi_id.into(),
-            point: Embedding::from(coi.embedding),
-            stats: CoiStats {
-                view_count: coi.view_count as usize,
-                view_time: Duration::from_millis(coi.view_time_ms as u64),
-                last_view: coi.last_view.into(),
-            },
-        })
-        .collect();
+        let interactions = interactions.into_iter();
+        if interactions.len() == 0 {
+            return Ok(());
+        }
+        //FIXME micro benchmark and chunking+persist abstraction
+        let persist = interactions.len() < 10;
+
+        let mut builder = QueryBuilder::new(
+            "INSERT INTO interaction (doc_id, user_id, time_stamp, user_reaction)",
+        );
+        let mut iter = interactions.peekable();
+        while iter.peek().is_some() {
+            let chunk = (&mut iter).take(Database::BIND_LIMIT / 4);
+            builder
+                .reset()
+                .push_values(chunk, |mut builder, (document_id, interaction)| {
+                    builder
+                        .push_bind(document_id)
+                        .push_bind(user_id)
+                        .push_bind(now)
+                        .push_bind(interaction as i16);
+                })
+                .push(
+                    "ON CONFLICT (doc_id, user_id, time_stamp) DO UPDATE SET
```

Comment:
is a white space required here?
```suggestion
                    " ON CONFLICT (doc_id, user_id, time_stamp) DO UPDATE SET
```